### ISSUE TITLE
Is there any plan to add kosmos-2 to the transformers. 

### ISSUE THREAD

**Issue by BIGBALLON** [huggingface/transformers#24671]
### Model description

Kosmos-2 is a grounded multimodal large language model, which integrates grounding and referring capabilities compared with Kosmos-1. The model can accept image regions selected by the user using bounding boxes as input, provide visual answers (i.e., bounding boxes), and ground the text output to the visual world.

**Is there any plan to add this model to the transformers.**

### Open source status

- [X] The model implementation is available
- [X] The model weights are available

### Provide useful links for the implementation

Code: https://github.com/microsoft/unilm/tree/master/kosmos-2
Paper: https://arxiv.org/abs/2306.14824
Weight: the checkpoint can be downloaded from [here](https://conversationhub.blob.core.windows.net/beit-share-public/kosmos-2/kosmos-2.pt?sv=2021-10-04&st=2023-06-08T11%3A16%3A02Z&se=2033-06-09T11%3A16%3A00Z&sr=c&sp=r&sig=N4pfCVmSeq4L4tS8QbrFVsX6f6q844eft8xSuXdxU48%3D)  
VQA demo: [here](https://github.com/BIGBALLON/kosmos-2-gd)

**Comment by ydshieh** [comment]
Thank you for mentioning this :-). There is some early discussion within the team. I will come back to you once we have some decision.

**Comment by ydshieh** [comment]
This is tracked in PR #24709. (so far empty, but I will try to 🚀 )

**Comment by BIGBALLON** [comment]
@ydshieh  I'm very excited to hear this news. I sincerely appreciate your efforts.

**Comment by Rajmehta123** [comment]
any updates?

**Comment by ydshieh** [comment]
Still on it (slowly) 🤗

**Comment by Rajmehta123** [comment]
Sure. Thank you. Appreciate those efforts.

**Comment by yolandalalala** [comment]
I just want to say a big thank you for your effort @ydshieh!

**Comment by BIGBALLON** [comment]
@Rajmehta123 @yolandalalala @vanpelt 

This [project](https://github.com/BIGBALLON/kosmos-2-gd) can be provided for everyone to try, I hope it can help everyone

**Comment by ydshieh** [comment]
Very nice! @BIGBALLON Thanks a lot!

**Comment by BIGBALLON** [comment]
@ydshieh Thank you again for your great contribution!

**Comment by hujunchao** [comment]
Amazing! @BIGBALLON Thanks a lot!

**Comment by ydshieh** [comment]
Just want to give a update: I am almost done the coding - just need to put everything together.

(The model might ends up as a custom code on the Hub instead of directly available in `transformers` - I am not sure)

**Comment by BIGBALLON** [comment]
Hi, @ydshieh , is there any update? 😄 

> Just want to give a update: I am almost done the coding - just need to put everything together to finalize.
> 
> (The model might ends up as a custom code on the Hub instead of directly available in `transformers` - I am not sure)

**Comment by ydshieh** [comment]
Hi @BIGBALLON 

Sorry for the delay due to some internal task! I will make it available on the Hub this week 🔥 .
(And we will see if it would be directly in `transformers` later).

**Comment by ydshieh** [comment]
Hi @BIGBALLON 

As promised, I put the code on this HuggingFace Hub repository [ydshieh/kosmos-2-patch14-224](https://huggingface.co/ydshieh/kosmos-2-patch14-224). You can use it like the code snippet at the end. I will give something like

> <grounding> An image of<phrase> a snowman</phrase><object><patch_index_0044><patch_index_0863></object> warming himself by<phrase> a fire</phrase><object><patch_index_0005><patch_index_0911></object>.

I will provide a more complete post-processing function though.

Note that if this model would be added into `transformers` codebase, there might be some changes which I could not guarantee it won't break the current behavior.

```python
from PIL import Image
from transformers import AutoProcessor, AutoModelForVision2Seq


model = AutoModelForVision2Seq.from_pretrained("ydshieh/kosmos-2-patch14-224", trust_remote_code=True)
processor = AutoProcessor.from_pretrained("ydshieh/kosmos-2-patch14-224", trust_remote_code=True)

prompt = "<grounding>An image of"
image = Image.open("snowman.jpg")

inputs = processor(text=prompt, images=image, return_tensors="pt")

generated_ids = model.generate(
    pixel_values=inputs["pixel_values"],
    input_ids=inputs["input_ids"][:, :-1],
    attention_mask=inputs["attention_mask"][:, :-1],
    img_features=None,
    img_attn_mask=inputs["img_attn_mask"][:, :-1],
    use_cache=True,
    max_new_tokens=64,
)
generated_text = processor.batch_decode(generated_ids, skip_special_tokens=True)[0]
result = processor.post_processor_generation(generated_text)
print(result)

```

**Comment by Rajmehta123** [comment]
Amazing work. Thank you.

**Comment by Rajmehta123** [comment]
Can this model be used for Q&A?

**Comment by ydshieh** [comment]
I am also trying to see what this model can do - in the paper, it can do more things than what the demo demonstrates

**Comment by BIGBALLON** [comment]
@ydshieh thanks again for your effort!!!

@Rajmehta123 VQA is supported only to change the prompts.

**Comment by ydshieh** [comment]
@BIGBALLON 

Yes, this model seems to be capable of doing quite different things, but it's challenging to showing this in a demo.
I am still looking what I can add, but please share your ideas too if any 🙏

**Comment by BIGBALLON** [comment]
Hi, @ydshieh , there are some suggestion: as for gradio app demo, we can provide two  outputs, `Text` and `Image` 

- for Visual Grounding task: use <grounding> with your grounding question for prompt, and the output image -> `Image`
- for VQA task:  do not use <grounding>, and only descript your question, then the output text -> `Text`

the keypoint is prompt, check this for more detials : https://github.com/BIGBALLON/kosmos-2-gd

**Comment by ydshieh** [comment]
Hi everyone!

I put a `Tasks` section in the README.md file

https://huggingface.co/ydshieh/kosmos-2-patch14-224/blob/main/README.md#tasks

@BIGBALLON For VQA, I also `<grounding>`, just like the official demo for image captioning uses `<grounding>`. It works well however.

**Comment by rabiulcste** [comment]
@ydshieh Any plan for supporting beam search in text-generaton?

**Comment by ydshieh** [comment]
The beam search is already supported in text-generaton for a long time. For this model, its default is beam size = 3.

**Comment by rabiulcste** [comment]
> The beam search is already supported in text-generaton for a long time. For this model, its default is beam size = 3.

I get this error while trying to use `num_beams = 3`. If set `use_cache=False`, the issue resolves, but the generation becomes 50x slower.

NotImplementedError: Make sure that a `_reorder_cache` function is correctly implemented in transformers_modules.ydshieh.kosmos-2-patch14-224.48e3edebaeb02dc9fe105f40e85a43a3b440dc72.modeling_kosmos2 to enable beam search for <class 'transformers_modules.ydshieh.kosmos-2-patch14-224.48e3edebaeb02dc9fe105f40e85a43a3b440dc72.modeling_kosmos2.Kosmos2TextForCausalLM'>

The full args for generation below.
```
model_name == "kosmos2":
          generated_ids = model.generate(
              pixel_values=batch["pixel_values"].to("cuda"),
              input_ids=batch["input_ids"][:, :-1].to("cuda"),
              attention_mask=batch["attention_mask"][:, :-1].to("cuda"),
              img_features=None,
              img_attn_mask=batch["img_attn_mask"][:, :-1].to("cuda"),
              max_new_tokens=args.max_length,
              length_penalty=args.length_penalty,
              num_beams=args.num_beams,
          )
```

**Comment by rabiulcste** [comment]
I copy and paste this snippet from `llama` to `kosmos2`. Now working fine.

https://github.com/huggingface/transformers/blob/62b20c9ecd6c9d2295265187a51ba0ea74ce046c/src/transformers/models/llama/modeling_llama.py#L901

**Comment by ydshieh** [comment]
Thank you for reporting. I will take a look, kinda strange here. Note there will be an official port in `transformers` soon, and my personal code on the Hub won't be the best place to use this model.

**Comment by rabiulcste** [comment]
> Thank you for reporting. I will take a look, kinda strange here. Note there will be an official port in `transformers` soon, and my personal code on the Hub won't be the best place to use this model.

Thanks, I will wait for the official support then!

**Comment by rabiulcste** [comment]
I'm getting an error on batch inference. 

```
  model = AutoModelForVision2Seq.from_pretrained(model_name_or_path, torch_dtype=autocast_dtype, device_map="auto")
  processor = AutoProcessor.from_pretrained(model_name_or_path)
  processor.tokenizer.padding_side = "left"
  
  ----
  encoding = processor(
  images=processed_batch["image"],
  text=processed_batch["prompted_question"],
  padding=True,
  return_tensors="pt",
)
```
```
  File "/path/site-packages/transformers/tokenization_utils_base.py", line 720, in as_tensor
    return torch.tensor(value)
ValueError: expected sequence of length 90 at dim 1 (got 89)

The above exception was the direct cause of the following exception:
-------
 raise ValueError(
ValueError: Unable to create tensor, you should probably activate truncation and/or padding with 'padding=True' 'truncation=True' to have batched tensors with the same length. Perhaps your features (`input_ids` in this case) have excessive nesting (inputs type `list` where type `int` is expected).
```

**Comment by ydshieh** [comment]
Hi @rabiulcste Please provide a complete code snippet. There are missing variable definitions above and I can't run it directly.

**Comment by NielsRogge** [comment]
KOSMOS-2 is available here: https://huggingface.co/docs/transformers/en/model_doc/kosmos-2
